lib/interactor/contract/contract.rb

```diff
@@ -0,0 +1,171 @@
+module Interactor
+  module Contract
+    VALID_TYPES = [:open, :closed]
+
+    def self.included(base)
+      base.class_eval do
+        extend ClassMethods
+      end
+
+      def violation_table
+        @violation_table ||= {}
+      end
+
+      def validate_contract_expectations
+        missing_properties.each do |property_name|
+          violation_table[property_name] = ContractError.new(self, missing: [property_name])
+        end
+
+        extra_properties.each do |property_name|
+          violation_table[property_name] = ContractError.new(self, undeclared: [property_name])
+        end
+      end
+
+      def missing_properties
+        @missing_properties ||= self.class.expected_properties.select do |attr|
+          !context.members.include?(attr)
+        end
+      end
+
+      def extra_properties
+        @extra_properties ||= begin
+          if self.class.contract_open?
+            []
+          else
+            context.members.select do |attr|
+              !self.class.expected_and_permitted_properties.include?(attr)
+            end
+          end
+        end
+      end
+
+      def ensure_contract_defaults
+        self.class.all_properties.each do |attr|
+          send(attr) if context[attr].nil? && self.class.default_for(attr)
+        end
+      end
+
+      def check_each_violation
+        return if violation_table.empty?
+        violation_table.each do |property_name, violation|
+          if block = self.class.property_table.get(property_name).on_violation
+            block.call(context)
+          elsif self.class.on_violation_block
+            self.class.on_violation_block.call(violation, property_name, context)
+          else
+            raise violation
+          end
+        end
+      end
+
+    end
+
+    module ClassMethods
+
+      # Core DSL
+      #
+      def contract_type(value)
+        raise "Invalid contract type '#{value}'" unless Interactor::Contract::VALID_TYPES.include?(value)
+        @contract_type = value
+      end
+
+      def property(attr, opts={}, &block)
+        opts.merge!(default: block) if block
+        property_table.set(attr, opts)
+
+        delegate_properties
+      end
+
+      def on_violation_for(*args, &block)
+        args.each do |arg|
+          if property = property_table.get(arg)
+            property.on_violation = block
+          else
+            property_table.set(arg, on_violation: block)
+          end
+        end
+      end
+
+      def on_violation(&block)
+        @on_violation_block = block
+      end
+
+      # Sugar for core DSL
+      def expects(*args, &block)
```

Comment:
There's a lot of repetition between `.expects`, `.permits`, and `.provides`. We could refactor into a helper method that passes in presence, I think.